targets/unity-v2/source/Shared/Public/Pipeline/OneDSEventPipeline.cs

```diff
@@ -0,0 +1,357 @@
+﻿﻿#if NET_4_6   
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using PlayFab.EventsModels;
+using PlayFab.Logger;
+using PlayFab.SharedModels;
+
+namespace PlayFab.Pipeline
+{
+    /// <summary>
+    /// The implementation of an event pipeline that manages work of multiple event processing stages
+    /// such as writings events, batching and sending over network.
+    /// The public methods of this class are meant to be thread-safe and can be called from multiple threads.
+    /// </summary>
+    public class OneDSEventPipeline : IEventPipeline, IDisposable
+    {
+        // A flag that indicates whether the pipeline is currently active (started)
+        // and accepts incoming data (events).
+        private volatile bool isActive;
+        private object isActiveLock = new object();
+
+        private OneDSEventPipelineSettings settings;
+
+        // External cancellation token is used to cancel the pipeline from outside code
+        private CancellationToken externalCancellationToken;
+
+        // Pipeline creates and uses its own internal cancellation token source (based on the external token)
+        // to be able to communicate a cancellation between its stages when the cancellation needs to happen
+        // inside the pipeline itself (for example, when an exception occurs). Cancellation in this case 
+        // is needed to unblock all stage threads from waiting on data in the buffers.
+        //
+        // More about C# pipelines, their cancellation and error handling see here:
+        // https://msdn.microsoft.com/en-us/library/ff963548.aspx
+        private CancellationTokenSource pipelineCancellationTokenSource;
+
+        // Buffers used in the pipeline
+        private BlockingCollection<IPlayFabEmitEventRequest> eventBuffer;
+        private BlockingCollection<TitleEventBatch> batchBuffer;
+        private BlockingCollection<PlayFabResult<WriteEventsResponse>> sendResultBuffer;
+
+        // Stages used in the pipeline
+        private IPipelineStage<IPlayFabEmitEventRequest, TitleEventBatch> batchingStage;
+        private IPipelineStage<TitleEventBatch, PlayFabResult<WriteEventsResponse>> sendingStage;
+
+        // references to individual stages of the pipeline task
+        private Task batchingTask; 
+        private Task sendingTask;
+
+        // The composition of all pipeline tasks including stages
+        private Task pipelineTask;
+
+        private ILogger logger;
+
+        public OneDSEventPipelineSettings Settings
```

Comment:
This could just be:
public OneDSEventPipelineSettings Settings { get; private set; }
It reduces complexity and more strongly indicates intent.
It could also be:
public readonly OneDSEventPipelineSettings Settings = new OneDSEventPipelineSettings();
Which would be even better, if that's a viable workflow.

Reply:
If it's needed to support runtime 3.5 and older version of C# please feel free to keep it. I'd rather have one codebase.

---
In reply to: [264020099](https://github.com/PlayFab/SDKGenerator/pull/436#discussion_r264020099) [](ancestors = 264020099)